src/tests/simulation/tests/integration_tests/sink/exactly_once_utils.rs

```diff
@@ -0,0 +1,915 @@
+            );
+            self.store.inc_err();
+            self.store.inner().err_events.push('2');
+            return Err(SinkError::Internal(anyhow::anyhow!(
+                "fail to commit to external sink."
+            )));
+        }
+        let file_ids = metadatas.into_iter().map(|metadata| {
+            let Metadata::Serialized(serialized) = metadata.metadata.unwrap();
+            usize::from_le_bytes((serialized.metadata.as_slice()).try_into().unwrap())
+        });
+        self.store.insert_many(
+            snapshot_id,
+            file_ids
+                .into_iter()
+                .map(|file_id| self.staging_store.get(file_id))
+                .flatten(),
+        );
+        println!("Succeeded to commit to mock iceberg table in epoch {epoch}.");
+
+        if thread_rng().random_ratio(self.err_rate_vec[3].load(Relaxed), u32::MAX) {
+            println!(
+                "Error injection point 3 -- Error occur after commit to external sink and before mark row deleted."
+            );
+            self.store.inc_err();
+            self.store.inner().err_events.push('3');
+            return Err(SinkError::Internal(anyhow::anyhow!(
+                "fail to mark row deleted."
+            )));
+        }
+
+        self.mark_row_is_committed_by_sink_id_and_end_epoch(
+            &self.db,
+            self.sink_id.sink_id(),
+            epoch,
+        )
+        .await?;
+        println!(
+            "Sink id = {}: succeeded mark pre commit metadata in epoch {} to deleted.",
+            self.sink_id, epoch
+        );
+
+        if thread_rng().random_ratio(self.err_rate_vec[4].load(Relaxed), u32::MAX) {
+            println!(
+                "Error injection point 4 -- Error occur after mark committed before deleting previous item."
+            );
+            self.store.inc_err();
+            self.store.inner().err_events.push('4');
+            return Err(SinkError::Internal(anyhow::anyhow!(
+                "fail to delete previous item in meta store."
+            )));
+        }
+
+        self.delete_row_by_sink_id_and_end_epoch(&self.db, self.sink_id.sink_id(), epoch)
+            .await?;
+        self.store.inner().err_events.push('x');
+        Ok(())
+    }
+
+    async fn re_commit(
+        &mut self,
+        epoch: u64,
+        metadata: Vec<SinkMetadata>,
+        snapshot_id: i64,
+    ) -> risingwave_connector::sink::Result<()> {
+        println!("Starting mock iceberg re commit in epoch {}.", epoch);
+        self.commit_inner(epoch, metadata, Some(snapshot_id))
+            .await?;
+
+        Ok(())
+    }
+}
+
+// mock iceberg sink handle system table
+impl SimulationTestIcebergCommitter {
+    async fn persist_pre_commit_metadata(
+        &self,
+        db: DatabaseConnection,
+        start_epoch: u64,
+        end_epoch: u64,
+        pre_commit_metadata: Vec<u8>,
+        snapshot_id: i64,
+    ) -> Result<()> {
+        let m = exactly_once_iceberg_sink::ActiveModel {
```

Comment:
It seems that we are still accessing the internal meta store table manually in the test. In this way, the logic of accessing meta store table in the original implementation is not tested at all. Can we possibly reuse these logic of accessing meta store table in the original implementation and only mock the access to iceberg? 

In our offline discussion, we once mentioned that we can extract all access to iceberg table to a common trait, and then we can mock accessing iceberg via implementing this trait.

Reply:
I functionalized the logic of the operating system table after enabling iceberg exactly once, and then used the same set of logic in the test and real environments.